File: python/src/kryon/runtime.py
"""The synchronous Kryon runtime.

Two operations, deliberately separate:

* :meth:`Runtime.execute` runs a program to completion and hands back everything it
  produced. Use it when you want the answer.
* :meth:`Runtime.spawn` starts a program and hands back a :class:`Process` you can write
  to, read from and signal while it runs. Use it when you want a conversation.

For ``async``/``await``, see :mod:`kryon.aio`.
"""

from __future__ import annotations

import contextlib
import queue
import subprocess
import threading
import time
from collections.abc import Iterator, Mapping, Sequence
from types import TracebackType
from typing import Any, Callable, Optional, Tuple, Type

from kryon import _core
from kryon._core import CHUNK, QUEUE_DEPTH, Buffer, Reason
from kryon.errors import InvalidArguments, UnsupportedPlatform
from kryon.model import ExecutionOptions, ExecutionResult, Stream, TerminationReason

__all__ = ["Process", "Runtime"]

_EOF = object()


class Runtime:
    """An execution context holding default options.

    A ``Runtime`` holds configuration, not state. It is safe to share across threads, and
    creating one is cheap enough that you can also just make a new one::

        runtime = Runtime(timeout=30, encoding="utf-8", cwd="/srv/app")
        result = runtime.execute("git", ["status", "--porcelain"])

    Any default can be overridden per call. ``env`` merges with the runtime's ``env``
    rather than replacing it; every other option is replaced.
    """

    def __init__(self, **defaults: Any) -> None:
        self._defaults = ExecutionOptions().merge(defaults)
        self._defaults.validate()

    @property
    def defaults(self) -> ExecutionOptions:
        """The options this runtime applies when a call does not override them."""
        return self._defaults

    def __repr__(self) -> str:  # pragma: no cover - cosmetic
        return f"Runtime(defaults={self._defaults!r})"

    # -- one-shot execution ------------------------------------------------

    def execute(
        self, executable: Any, arguments: Optional[Sequence[Any]] = None, **options: Any
    ) -> ExecutionResult:
        """Run ``executable`` with ``arguments`` and return what happened.

        Arguments are passed to the operating system as a vector. Nothing in them is
        interpreted -- ``execute("echo", ["$HOME && rm -rf /"])`` prints that text
        literally. For shell semantics you must ask for them by name; see
        :meth:`execute_shell`.

        A process that could not be started raises. A process that started and then
        failed is returned, because a non-zero exit is information: ``grep`` exits ``1``
        to mean "no match". Pass ``check=True`` to raise on those too.

        :raises CommandNotFound: the executable could not be resolved
        :raises PermissionDenied: the executable could not be run
        :raises ProcessStartFailed: the process could not be created
        :raises InvalidArguments: the request was malformed
        """
        return self._execute(executable, arguments, options, shell=False)

    def execute_shell(self, command_line: str, **options: Any) -> ExecutionResult:
        """Run ``command_line`` through the system shell.

        .. warning::
           The shell interprets quoting, globbing, variable expansion, pipes and command
           chaining. Building this string from untrusted input is a command-injection
           vulnerability. If you are interpolating a value, you almost certainly want
           :meth:`execute` with an argument list instead.

        The shell is ``/bin/sh -c`` on POSIX and ``%COMSPEC% /c`` on Windows.

        This method exists as a separate name on purpose. A ``shell=True`` flag sitting
        among a dozen options is easy to set by accident and easy to miss in review; a
        differently named method is not.
        """
        if not isinstance(command_line, str):
            raise InvalidArguments(
                f"command_line must be a string, got {type(command_line).__name__}"
            )
        return self._execute(command_line, None, options, shell=True)

    def _execute(
        self,
        executable: Any,
        arguments: Optional[Sequence[Any]],
        overrides: Mapping[str, Any],
        shell: bool,
    ) -> ExecutionResult:
        opts = self._defaults.merge(overrides)
        opts.validate()
        exe, args = _core.normalise(executable, arguments)
        _core.check_cwd(opts.cwd)

        started = time.monotonic()
        proc = _start(exe, args, opts, shell)
        reason = Reason()
        out = Buffer(opts.max_output_bytes)
        err = Buffer(opts.max_output_bytes)

        workers = [
            _thread(self._pump, proc.stdout, out, reason, proc, opts),
            _thread(self._pump, proc.stderr, err, reason, proc, opts),
            _thread(_feed, proc.stdin, _encode_stdin(opts)),
        ]
        try:
            proc.wait(timeout=opts.timeout)
        except subprocess.TimeoutExpired:
            reason.set(TerminationReason.TIMEOUT)
            _core.stop(proc, opts.kill_grace)
            proc.wait()
        except BaseException:
            # KeyboardInterrupt, or anything else unwinding the stack. Returning control
            # to the caller while leaving the child running is not an option.
            reason.set(TerminationReason.CANCELLED)
            _core.stop(proc, opts.kill_grace)
            raise
        finally:
            # A grandchild can hold the pipes open after the child exits, so this join is
            # bounded. Without the bound, `sh -c "sleep 100 &"` would hang the caller.
            for worker in workers:
                worker.join(timeout=max(opts.kill_grace, 1.0))
            _close(proc)

        termination, exit_code, sig = _core.classify(proc.returncode, reason.value)
        result = ExecutionResult(
            executable=exe,
            arguments=tuple(args),
            exit_code=exit_code,
            signal=sig,
            stdout=_core.decode(out.value(), opts.encoding),
            stderr=_core.decode(err.value(), opts.encoding),
            duration=time.monotonic() - started,
            termination=termination,
            pid=proc.pid,
            stdout_truncated=out.truncated,
            stderr_truncated=err.truncated,
        )
        return result.check() if opts.check else result

    @staticmethod
    def _pump(
        fh: Any,
        buffer: Buffer,
        reason: Reason,
        proc: "subprocess.Popen[bytes]",
        opts: ExecutionOptions,
    ) -> None:
        try:
            while True:
                data = fh.read1(CHUNK)
                if not data:
                    return
                if buffer.add(data) and reason.set(TerminationReason.OUTPUT_LIMIT):
                    _core.stop(proc, opts.kill_grace)
        except (OSError, ValueError):
            return  # pipe closed underneath us during shutdown

    # -- long-lived processes ----------------------------------------------

    def spawn(
        self, executable: Any, arguments: Optional[Sequence[Any]] = None, **options: Any
    ) -> Process:
        """Start ``executable`` and return a :class:`Process` to interact with.

        Returns as soon as the process has started. Use it as a context manager so the
        process cannot outlive the block::

            with runtime.spawn("python", ["-u", "worker.py"]) as proc:
                proc.write("job-1\\n")
                for stream, chunk in proc:
                    print(chunk)
        """
        opts = self._defaults.merge(options)
        opts.validate()
        exe, args = _core.normalise(executable, arguments)
        _core.check_cwd(opts.cwd)
        return Process(_start(exe, args, opts, shell=False), exe, args, opts)


class Process:
    """A running process you can talk to.

    Output arrives through :attr:`output` as ``(Stream, bytes)`` pairs in the order
    Kryon observed them. Chunk boundaries mean nothing -- they reflect how the operating
    system delivered the data, not lines or records.

    The stream is bounded. If you stop consuming, Kryon stops reading, the pipe fills and
    the child blocks. That is backpressure working, not a hang: a program that produces
    faster than you consume must be slowed down somewhere, and doing it in the kernel
    buffer is better than doing it in your heap.

    Construct via :meth:`Runtime.spawn`, not directly.
    """

    def __init__(
        self,
        proc: subprocess.Popen[bytes],
        executable: str,
        arguments: Sequence[str],
        options: ExecutionOptions,
    ) -> None:
        self._proc = proc
        self._executable = executable
        self._arguments = tuple(arguments)
        self._options = options
        self._started = time.monotonic()
        self._reason = Reason()
        self._queue: queue.Queue[Any] = queue.Queue(maxsize=QUEUE_DEPTH)
        self._consumed = False
        self._bytes_seen = 0
        self._closed = False
        self._readers = [
            _thread(self._read, Stream.STDOUT, proc.stdout),
            _thread(self._read, Stream.STDERR, proc.stderr),
        ]
        self._timer: Optional[threading.Timer] = None
        if options.timeout is not None:
            self._timer = threading.Timer(options.timeout, self._on_timeout)
            self._timer.daemon = True
            self._timer.start()

    # -- identity ----------------------------------------------------------

    @property
    def pid(self) -> int:
        """The operating-system process id."""
        return self._proc.pid

    @property
    def running(self) -> bool:
        """Whether the process is still alive."""
        return self._proc.poll() is None

    @property
    def exit_code(self) -> Optional[int]:
        """The exit status once the process has been reaped, otherwise ``None``."""
        return self._proc.poll()

    def __repr__(self) -> str:  # pragma: no cover - cosmetic
        state = "running" if self.running else f"exited {self.exit_code}"
        return f"Process(pid={self.pid}, {self._executable!r}, {state})"

    # -- input -------------------------------------------------------------

    def write(self, data: Any) -> None:
        """Write to the child's stdin.

        Text is encoded with the runtime's ``encoding``, or UTF-8 when none is set.
        Raises if stdin is already closed -- dropping input silently is the failure mode
        that produces hangs nobody can reproduce.
        """
        if self._proc.stdin is None or self._proc.stdin.closed:
            raise ValueError(f"stdin of pid {self.pid} is closed")
        if isinstance(data, str):
            data = data.encode(self._options.encoding or "utf-8")
        try:
            self._proc.stdin.write(data)
            self._proc.stdin.flush()
        except BrokenPipeError:
            # The child stopped reading. Its exit code, not this write, is the story.
            pass

    def close_stdin(self) -> None:
        """Close stdin, signalling end-of-input to the child."""
        if self._proc.stdin is not None and not self._proc.stdin.closed:
            with contextlib.suppress(OSError):
                self._proc.stdin.close()

    # -- output ------------------------------------------------------------

    @property
    def output(self) -> Iterator[Tuple[Stream, bytes]]:
        """Iterate ``(stream, chunk)`` pairs until both pipes reach end-of-input.

        There is one consumer. Iterating twice raises, because the second iterator would
        silently steal chunks from the first.
        """
        if self._consumed:
            raise ValueError(
                f"output of pid {self.pid} is already being consumed; a process has one "
                "output stream, and two readers would each get an arbitrary half of it"
            )
        self._consumed = True
        return self._iterate()

    def __iter__(self) -> Iterator[Tuple[Stream, bytes]]:
        return self.output

    def _iterate(self) -> Iterator[Tuple[Stream, bytes]]:
        open_streams = 2
        while open_streams:
            item = self._queue.get()
            if item is _EOF:
                open_streams -= 1
                continue
            yield item

    def _read(self, stream: Stream, fh: Any) -> None:
        try:
            while True:
                data = fh.read1(CHUNK)
                if not data:
                    return
                self._bytes_seen += len(data)
                limit = self._options.max_output_bytes
                if (
                    limit is not None
                    and self._bytes_seen > limit
                    and self._reason.set(TerminationReason.OUTPUT_LIMIT)
                ):
                    _core.stop(self._proc, self._options.kill_grace)
                self._queue.put((stream, data))
        except (OSError, ValueError):
            return
        finally:
            self._queue.put(_EOF)

    # -- lifecycle ---------------------------------------------------------

    def signal(self, number: int) -> None:
        """Send a specific signal to the process.

        :raises UnsupportedPlatform: on Windows, which has no signals to send. Use
            :meth:`terminate` there, and know that it does not give the child a chance
            to clean up.
        """
        if _core.WINDOWS:
            raise UnsupportedPlatform(
                "Windows has no signals; use terminate(), which kills the process "
                "outright without letting it clean up"
            )
        self._proc.send_signal(number)

    def terminate(self) -> None:
        """Request a polite stop: ``SIGTERM`` on POSIX, ``TerminateProcess`` on Windows.

        On Windows this is identical to :meth:`kill`. There is no graceful stop.
        """
        with contextlib.suppress(OSError):
            self._proc.terminate()

    def kill(self) -> None:
        """Force a stop: ``SIGKILL`` on POSIX, ``TerminateProcess`` on Windows."""
        with contextlib.suppress(OSError):
            self._proc.kill()

    def wait(self, timeout: Optional[float] = None) -> ExecutionResult:
        """Block until the process exits and return the outcome.

        ``stdout`` and ``stderr`` on the result are empty: the output was streamed to
        you through :attr:`output` and is deliberately not buffered a second time.

        :raises ProcessTimeout: ``timeout`` elapsed. The process is left running -- this
            is a wait, not a stop. Call :meth:`terminate` or :meth:`close` for that.
        """
        try:
            self._proc.wait(timeout=timeout)
        except subprocess.TimeoutExpired:
            from kryon.errors import ProcessTimeout

            raise ProcessTimeout(
                f"{self._executable!r} (pid {self.pid}) still running after {timeout}s"
            ) from None
        if self._timer is not None:
            self._timer.cancel()
        termination, exit_code, sig = _core.classify(self._proc.returncode, self._reason.value)
        empty = "" if self._options.encoding else b""
        return ExecutionResult(
            executable=self._executable,
            arguments=self._arguments,
            exit_code=exit_code,
            signal=sig,
            stdout=empty,
            stderr=empty,
            duration=time.monotonic() - self._started,
            termination=termination,
            pid=self._proc.pid,
        )

    def close(self) -> None:
        """Terminate the process if it is still running and release every resource.

        Idempotent. This is what leaving a ``with`` block does.
        """
        if self._closed:
            return
        self._closed = True
        if self._timer is not None:
            self._timer.cancel()
        self.close_stdin()
        if self.running:
            self._reason.set(TerminationReason.CANCELLED)
            _core.stop(self._proc, self._options.kill_grace)
        # A reader blocked on a full queue will never see the pipe close, so drain until
        # the threads finish. Without this, close() on an unconsumed process leaks two
        # threads and two file descriptors.
        deadline = time.monotonic() + max(self._options.kill_grace, 1.0)
        while any(t.is_alive() for t in self._readers) and time.monotonic() < deadline:
            with contextlib.suppress(queue.Empty):
                self._queue.get(timeout=0.05)
        for reader in self._readers:
            reader.join(timeout=0.5)
        _close(self._proc)
        with contextlib.suppress(subprocess.TimeoutExpired):  # pragma: no cover
            self._proc.wait(timeout=1.0)

    def _on_timeout(self) -> None:
        if self.running and self._reason.set(TerminationReason.TIMEOUT):
            _core.stop(self._proc, self._options.kill_grace)

    def __enter__(self) -> Process:
        return self

    def __exit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc: Optional[BaseException],
        tb: Optional[TracebackType],
    ) -> None:
        self.close()

    def __del__(self) -> None:  # pragma: no cover - interpreter shutdown ordering
        if not self._closed and self._proc.poll() is None:
            import warnings

            warnings.warn(
                f"Process(pid={self._proc.pid}, {self._executable!r}) was garbage "
                "collected while still running; use it as a context manager or call "
                "close()",
                ResourceWarning,
                stacklevel=2,
            )
            self.close()


# -- module-private helpers ------------------------------------------------


def _start(
    executable: str, arguments: Sequence[str], opts: ExecutionOptions, shell: bool
) -> subprocess.Popen[bytes]:
    try:
        return subprocess.Popen(
            _core.popen_args(executable, arguments, shell),
            shell=shell,
            cwd=opts.cwd,
            env=_core.build_env(opts),
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
    except OSError as exc:
        raise _core.map_start_error(exc, executable, opts.cwd) from exc


def _encode_stdin(opts: ExecutionOptions) -> bytes:
    data = opts.stdin
    if data is None:
        return b""
    if isinstance(data, str):
        return data.encode(opts.encoding or "utf-8")
    return bytes(data)


def _feed(fh: Any, data: bytes) -> None:
    try:
        if data:
            fh.write(data)
            fh.flush()
    except (OSError, ValueError):
        pass  # child exited before reading its input; its exit code is the story
    finally:
        with contextlib.suppress(OSError, ValueError):
            fh.close()


def _thread(target: Callable[..., None], *args: Any) -> threading.Thread:
    thread = threading.Thread(target=target, args=args, daemon=True)
    thread.start()
    return thread


def _close(proc: subprocess.Popen[bytes]) -> None:
    for fh in (proc.stdin, proc.stdout, proc.stderr):
        if fh is not None and not fh.closed:
            with contextlib.suppress(OSError, ValueError):
                fh.close()

File: java/src/main/java/io/github/piyushmishra00/kryon/Runtime.java
package io.github.piyushmishra00.kryon;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

/**
 * The Kryon runtime.
 *
 * <p>Two operations, deliberately separate:
 *
 * <ul>
 *   <li>{@link #execute} runs a program to completion and hands back everything it produced. Use
 *       it when you want the answer.
 *   <li>{@link #spawn} starts a program and hands back a {@link KryonProcess} you can write to,
 *       read from and signal while it runs. Use it when you want a conversation.
 * </ul>
 *
 * <p>A {@code Runtime} holds configuration, not state. It is safe to share across threads, and
 * creating one is cheap enough that you can also just make a new one:
 *
 * <pre>{@code
 * Runtime runtime = new Runtime(ExecutionOptions.builder()
 *         .charset(StandardCharsets.UTF_8)
 *         .timeout(Duration.ofSeconds(30))
 *         .build());
 *
 * ExecutionResult result = runtime.execute("git", List.of("status", "--porcelain"));
 * }</pre>
 */
public final class Runtime {

    static final boolean WINDOWS =
            System.getProperty("os.name", "").toLowerCase(java.util.Locale.ROOT).contains("win");

    /**
     * Preserved even when {@code clearEnv} is set. Windows binaries -- including the ones in
     * {@code System32} -- routinely fail to start without these.
     */
    private static final List<String> WINDOWS_ESSENTIAL = List.of("SystemRoot", "SystemDrive");

    private static final int CHUNK = 65536;

    private final ExecutionOptions defaults;

    /** Creates a runtime with no default options. */
    public Runtime() {
        this(ExecutionOptions.defaults());
    }

    /**
     * Creates a runtime with default options applied to every call.
     *
     * @param defaults the defaults; each call may override them
     */
    public Runtime(ExecutionOptions defaults) {
        this.defaults = Objects.requireNonNull(defaults, "defaults");
        this.defaults.validate();
    }

    /**
     * The options this runtime applies when a call does not override them.
     *
     * @return the defaults
     */
    public ExecutionOptions defaults() {
        return defaults;
    }

    /**
     * Runs {@code executable} with no arguments.
     *
     * @param executable the program to run
     * @return what happened
     */
    public ExecutionResult execute(String executable) {
        return execute(executable, List.of(), null);
    }

    /**
     * Runs {@code executable} with {@code arguments}.
     *
     * @param executable the program to run
     * @param arguments the argument vector
     * @return what happened
     */
    public ExecutionResult execute(String executable, List<String> arguments) {
        return execute(executable, arguments, null);
    }

    /**
     * Runs {@code executable} with {@code arguments} and returns what happened.
     *
     * <p>Arguments are passed to the operating system as a vector. Nothing in them is interpreted
     * -- {@code execute("echo", List.of("$HOME && rm -rf /"))} prints that text literally. For
     * shell semantics you must ask for them by name; see {@link #executeShell}.
     *
     * <p>A process that could not be started throws. A process that started and then failed is
     * returned, because a non-zero exit is information: {@code grep} exits {@code 1} to mean "no
     * match". Set {@code check(true)} to throw on those too.
     *
     * @param executable the program to run
     * @param arguments the argument vector
     * @param options per-call options layered over this runtime's defaults, may be null
     * @return what happened
     * @throws CommandNotFoundException if the executable could not be resolved
     * @throws PermissionDeniedException if the executable could not be run
     * @throws ProcessStartFailedException if the process could not be created
     * @throws InvalidArgumentsException if the request was malformed
     */
    public ExecutionResult execute(
            String executable, List<String> arguments, ExecutionOptions options) {
        return run(executable, arguments, options, false);
    }

    /**
     * Runs {@code commandLine} through the system shell.
     *
     * <p><strong>The shell interprets quoting, globbing, variable expansion, pipes and command
     * chaining. Building this string from untrusted input is a command-injection
     * vulnerability.</strong> If you are interpolating a value, you almost certainly want
     * {@link #execute} with an argument list instead.
     *
     * <p>The shell is {@code /bin/sh -c} on POSIX and {@code cmd.exe /d /s /c} on Windows.
     *
     * <p>This method exists as a separate name on purpose. A {@code shell(true)} flag sitting
     * among a dozen options is easy to set by accident and easy to miss in review; a differently
     * named method is not.
     *
     * @param commandLine the command line, interpreted by the shell
     * @return what happened
     */
    public ExecutionResult executeShell(String commandLine) {
        return executeShell(commandLine, null);
    }

    /**
     * Runs {@code commandLine} through the system shell.
     *
     * @param commandLine the command line, interpreted by the shell
     * @param options per-call options layered over this runtime's defaults, may be null
     * @return what happened
     * @see #executeShell(String)
     */
    public ExecutionResult executeShell(String commandLine, ExecutionOptions options) {
        if (commandLine == null) {
            throw new InvalidArgumentsException("commandLine must not be null");
        }
        return run(commandLine, List.of(), options, true);
    }

    private ExecutionResult run(
            String executable, List<String> arguments, ExecutionOptions overrides, boolean shell) {

        ExecutionOptions options = defaults.mergedWith(overrides);
        options.validate();
        List<String> argv = normalise(executable, arguments);
        checkCwd(options);

        long started = System.nanoTime();
        Process process = start(executable, argv, options, shell);

        AtomicReference<TerminationReason> reason = new AtomicReference<>();
        Sink out = new Sink(options.maxOutputBytes().orElse(null));
        Sink err = new Sink(options.maxOutputBytes().orElse(null));

        Runnable stopper = () -> stop(process, options.killGrace());

        Thread pumpOut = pump(process.getInputStream(), out, reason, stopper);
        Thread pumpErr = pump(process.getErrorStream(), err, reason, stopper);
        Thread feeder = feed(process, options);

        boolean exited;
        try {
            if (options.timeout().isPresent()) {
                Duration timeout = options.timeout().get();
                exited = process.waitFor(timeout.toNanos(), TimeUnit.NANOSECONDS);
                if (!exited) {
                    reason.compareAndSet(null, TerminationReason.TIMEOUT);
                    stopper.run();
                    process.waitFor();
                }
            } else {
                process.waitFor();
            }
        } catch (InterruptedException interrupted) {
            // Returning control while leaving the child running is not an option.
            reason.compareAndSet(null, TerminationReason.CANCELLED);
            stopper.run();
            Thread.currentThread().interrupt();
            throw new ProcessCancelledException(
                    "'" + executable + "' was interrupted while running", null);
        } finally {
            // Bounded, because a grandchild can hold the pipes open after the child exits.
            long graceMillis = Math.max(options.killGrace().toMillis(), 1000L);
            join(pumpOut, graceMillis);
            join(pumpErr, graceMillis);
            join(feeder, graceMillis);
        }

        Outcome outcome = classify(process, reason.get());
        ExecutionResult result =
                new ExecutionResult(
                        executable,
                        argv,
                        outcome.exitCode(),
                        outcome.signal(),
                        out.value(),
                        err.value(),
                        options.charset().orElse(null),
                        Duration.ofNanos(System.nanoTime() - started),
                        outcome.termination(),
                        process.pid(),
                        out.truncated(),
                        err.truncated());

        return options.check() ? result.checked() : result;
    }

    /**
     * Starts {@code executable} and returns a {@link KryonProcess} to interact with.
     *
     * <p>Returns as soon as the process has started. {@code KryonProcess} is
     * {@link AutoCloseable}, so use try-with-resources and the process cannot outlive the block.
     *
     * @param executable the program to run
     * @param arguments the argument vector
     * @return the running process
     */
    public KryonProcess spawn(String executable, List<String> arguments) {
        return spawn(executable, arguments, null);
    }

    /**
     * Starts {@code executable} and returns a {@link KryonProcess} to interact with.
     *
     * @param executable the program to run
     * @param arguments the argument vector
     * @param options per-call options layered over this runtime's defaults, may be null
     * @return the running process
     */
    public KryonProcess spawn(
            String executable, List<String> arguments, ExecutionOptions options) {
        ExecutionOptions merged = defaults.mergedWith(options);
        merged.validate();
        List<String> argv = normalise(executable, arguments);
        checkCwd(merged);
        return new KryonProcess(start(executable, argv, merged, false), executable, argv, merged);
    }

    // ------------------------------------------------------------------ internals

    private static List<String> normalise(String executable, List<String> arguments) {
        if (executable == null) {
            throw new InvalidArgumentsException("executable must not be null");
        }
        if (executable.isEmpty()) {
            throw new InvalidArgumentsException("executable must not be empty");
        }
        List<String> argv = new ArrayList<>();
        List<String> source = arguments == null ? List.of() : arguments;
        for (int i = 0; i < source.size(); i++) {
            String value = source.get(i);
            if (value == null) {
                throw new InvalidArgumentsException(
                        "argument "
                                + i
                                + " must not be null; Kryon does not stringify arguments for you, "
                                + "because guessing how to render a value into a command line is "
                                + "how injection bugs start");
            }
            argv.add(value);
        }
        return argv;
    }

    private static void checkCwd(ExecutionOptions options) {
        if (options.cwd().isEmpty()) {
            return;
        }
        Path cwd = options.cwd().get();
        if (!Files.exists(cwd)) {
            throw new ProcessStartFailedException("working directory does not exist: " + cwd);
        }
        if (!Files.isDirectory(cwd)) {
            throw new ProcessStartFailedException("working directory is not a directory: " + cwd);
        }
    }

    private static Process start(
            String executable, List<String> argv, ExecutionOptions options, boolean shell) {

        List<String> command = new ArrayList<>();
        if (shell) {
            if (WINDOWS) {
                command.addAll(List.of(shellExecutable(), "/d", "/s", "/c", executable));
            } else {
                command.addAll(List.of(shellExecutable(), "-c", executable));
            }
        } else {
            command.add(executable);
            command.addAll(argv);
        }

        ProcessBuilder builder = new ProcessBuilder(command);
        options.cwd().ifPresent(cwd -> builder.directory(cwd.toFile()));
        applyEnvironment(builder, options);

        try {
            return builder.start();
        } catch (IOException error) {
            throw mapStartError(error, executable);
        }
    }

    private static String shellExecutable() {
        if (WINDOWS) {
            String comspec = System.getenv("COMSPEC");
            return comspec == null || comspec.isEmpty() ? "cmd.exe" : comspec;
        }
        return "/bin/sh";
    }

    private static void applyEnvironment(ProcessBuilder builder, ExecutionOptions options) {
        Map<String, String> environment = builder.environment();
        if (options.clearEnv()) {
            Map<String, String> preserved = new HashMap<>();
            if (WINDOWS) {
                for (String key : WINDOWS_ESSENTIAL) {
                    String value = environment.get(key);
                    if (value != null) {
                        preserved.put(key, value);
                    }
                }
            }
            environment.clear();
            environment.putAll(preserved);
        }
        options.env()
                .forEach(
                        (key, value) -> {
                            if (value == null) {
                                environment.remove(key);
                            } else {
                                environment.put(key, value);
                            }
                        });
    }

    private static KryonException mapStartError(IOException error, String executable) {
        String message = String.valueOf(error.getMessage());
        String lower = message.toLowerCase(java.util.Locale.ROOT);
        if (lower.contains("no such file")
                || lower.contains("cannot find")
                || lower.contains("cannot run program")
                        && (lower.contains("error=2") || lower.contains("error=3"))) {
            return new CommandNotFoundException("executable not found: '" + executable + "'", error);
        }
        if (lower.contains("permission denied")
                || lower.contains("access is denied")
                || lower.contains("error=13")) {
            return new PermissionDeniedException(
                    "not permitted to execute '" + executable + "': " + message, error);
        }
        return new ProcessStartFailedException(
                "could not start '" + executable + "': " + message, error);
    }

    private static Thread pump(
            InputStream source,
            Sink sink,
            AtomicReference<TerminationReason> reason,
            Runnable stopper) {

        Thread thread =
                new Thread(
                        () -> {
                            byte[] buffer = new byte[CHUNK];
                            try {
                                int read;
                                while ((read = source.read(buffer)) != -1) {
                                    if (sink.add(buffer, read)
                                            && reason.compareAndSet(
                                                    null, TerminationReason.OUTPUT_LIMIT)) {
                                        stopper.run();
                                    }
                                }
                            } catch (IOException ignored) {
                                // Pipe closed underneath us during shutdown.
                            }
                        },
                        "kryon-pump");
        thread.setDaemon(true);
        thread.start();
        return thread;
    }

    private static Thread feed(Process process, ExecutionOptions options) {
        byte[] data = options.stdin().orElse(new byte[0]);
        Thread thread =
                new Thread(
                        () -> {
                            try (var stdin = process.getOutputStream()) {
                                if (data.length > 0) {
                                    stdin.write(data);
                                    stdin.flush();
                                }
                            } catch (IOException ignored) {
                                // The child exited before reading its input; its exit code is the
                                // story.
                            }
                        },
                        "kryon-feed");
        thread.setDaemon(true);
        thread.start();
        return thread;
    }

    private static void join(Thread thread, long millis) {
        try {
            thread.join(millis);
        } catch (InterruptedException interrupted) {
            Thread.currentThread().interrupt();
        }
    }

    /**
     * Terminates politely, then kills. The single termination path in this SDK.
     *
     * <p>On Windows both steps are {@code TerminateProcess}: there is no graceful stop, and the
     * child gets no chance to flush.
     */
    static void stop(Process process, Duration killGrace) {
        if (!process.isAlive()) {
            return;
        }
        process.destroy();
        try {
            if (process.waitFor(killGrace.toMillis(), TimeUnit.MILLISECONDS)) {
                return;
            }
        } catch (InterruptedException interrupted) {
            Thread.currentThread().interrupt();
        }
        process.destroyForcibly();
        try {
            process.waitFor(killGrace.toMillis(), TimeUnit.MILLISECONDS);
        } catch (InterruptedException interrupted) {
            Thread.currentThread().interrupt();
        }
    }

    /** The reported outcome: what Kryon decided, plus the raw exit status. */
    record Outcome(TerminationReason termination, Integer exitCode, Integer signal) {}

    /**
     * Maps a raw exit value plus any Kryon intervention to the reported outcome.
     *
     * <p>The JVM reports a signal death as {@code 128 + signum} on POSIX and gives no signal
     * number of its own, which is why {@code signal} is derived rather than read.
     */
    static Outcome classify(Process process, TerminationReason reason) {
        int raw = process.exitValue();
        boolean signaled = !WINDOWS && raw > 128 && raw < 192;
        TerminationReason natural =
                signaled ? TerminationReason.SIGNALED : TerminationReason.EXITED;
        return new Outcome(
                reason == null ? natural : reason,
                signaled ? null : raw,
                signaled ? raw - 128 : null);
    }

    /** A byte sink that stops growing at a limit and remembers that it did. */
    static final class Sink {

        private final Long limit;
        private final ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        private volatile boolean truncated;

        Sink(Long limit) {
            this.limit = limit;
        }

        /**
         * Appends {@code length} bytes. Returns true when the limit has just been exceeded.
         *
         * <p>Data past the limit is dropped rather than counted: the point of a cap is not to hold
         * the bytes.
         */
        synchronized boolean add(byte[] data, int length) {
            if (limit == null) {
                buffer.write(data, 0, length);
                return false;
            }
            long room = limit - buffer.size();
            if (room > 0) {
                buffer.write(data, 0, (int) Math.min(room, length));
            }
            if (length > room) {
                truncated = true;
                return true;
            }
            return false;
        }

        synchronized byte[] value() {
            return buffer.toByteArray();
        }

        boolean truncated() {
            return truncated;
        }
    }
}
